module/VuFindTheme/src/VuFindTheme/View/Helper/Slot.php

```diff
@@ -0,0 +1,209 @@
+     */
+    protected $stack = [];
+
+    /**
+     * Get the Slot instance. Create if instance doesn't exist.
+     *
+     * @param string $name Name of target block for action
+     *
+     * @return Slot
+     */
+    public function __invoke($name)
+    {
+        $this->stack[] = $name;
+        return $this;
+    }
+
+    /**
+     * Helper function to return blocks with prepends and appends
+     *
+     * @param string $name Name of target block for action
+     *
+     * @return string|any
+     */
+    protected function build($name)
+    {
+        $pre = $this->blockPrepends[$name] ?? [];
+        $block = $this->blocks[$name] ?? '';
+        $post = $this->blockAppends[$name] ?? [];
+        if (!empty($pre) || !empty($post)) {
+            // TODO: Warn about non-string blocks
+            return implode(' ', $pre) . $block . implode(' ', $post);
+        }
+        if (empty($block)) {
+            return null;
+        }
+        return $block;
+    }
+
+    /**
+     * Get current value of slot. Returns null if unset.
+     *
+     * @return string|null
+     */
+    public function get()
+    {
+        $name = array_pop($this->stack);
+        return $this->build($name);
+    }
+
+    /**
+     * Set current value of slot but only if unset.
+     *
+     * @param any $value Value to override if unset
+     *
+     * @return string|null
+     */
+    public function set($value)
+    {
+        $name = array_pop($this->stack);
+        if (!isset($this->blocks[$name])) {
+            $this->blocks[$name] = $value;
+        }
+        return $this->build($name);
+    }
+
+    /**
+     * Add string to list of block prepends.
+     *
+     * @param string $value Value to override if unset
+     *
+     * @return string
+     */
+    public function prepend($value)
+    {
+        $name = array_pop($this->stack);
+        if (!isset($this->blockPrepends[$name])) {
+            $this->blockPrepends[$name] = [$value];
+        } else {
+            array_unshift($this->blockPrepends[$name], $value);
+        }
+        return $this->build($name);
+    }
+
+    /**
+     * Add string to list of block appends.
+     *
+     * @param string $value Value to override if unset
+     *
+     * @return string
+     */
+    public function append($value)
+    {
+        $name = array_pop($this->stack);
+        if (!isset($this->blockAppends[$name])) {
+            $this->blockAppends[$name] = [$value];
+        } else {
+            array_push($this->blockAppends[$name], $value);
+        }
+        return $this->build($name);
+    }
+
+    /**
+     * Starts a buffer capture to override the value of a block.
+     *
+     * @return void
+     */
+    public function start()
+    {
+        array_pop($this->stack);
+        ob_start();
+    }
+
+    /**
+     * End a buffer capture to override the value of a block. Returns slot value.
+     *
+     * @param string $method SET/PREPEND/APPEND for where this buffer should be saved
```

Comment:
Might be nice to define constants for these inside the class.

Reply:
Based on my understanding I disagree. Setting constants would change the syntax from
```php
$this->slot('foo')->end('APPEND');
```
to something like
```php
$this->slot('foo')->end($this->slot::APPEND);
```
wouldn't it? Not so bad, I guess.

Reply:
If you defined the constants to the current string values, then the end user could decide whether to use the constant value or the string value. It just makes the internal code look a little more formal. Some of the Zend view helpers (like the header-related asset manipulation ones) do this sort of thing, if memory serves.